subsys/usb/usb_c/usbc_pe_common.c

```diff
@@ -0,0 +1,718 @@
+		return true;
+	}
+
+	return false;
+}
+
+/**
+ * @brief Check if a specific data message was received
+ */
+bool received_data_message(const struct device *dev,
+				  const union pd_header header,
+				  const enum pd_data_msg_type mt)
+{
+	struct usbc_port_data *data = dev->data;
+	struct protocol_layer_rx_t *prl_rx = data->prl_rx;
+
+	if (prl_rx->emsg.len > 0 &&
+	    header.message_type == mt &&
+	    header.extended == 0) {
+		return true;
+	}
+
+	return false;
+}
+
+/**
+ * @brief Check a DPM policy
+ */
+bool policy_check(const struct device *dev,
+			 const enum usbc_policy_check_t pc)
+{
+	struct usbc_port_data *data = dev->data;
+
+	if (data->policy_cb_check) {
+		return data->policy_cb_check(dev, pc);
+	} else {
+		return false;
+	}
+}
+
+/**
+ * @brief Notify the DPM of a policy change
+ */
+void policy_notify(const struct device *dev,
+			  const enum usbc_policy_notify_t notify)
+{
+	struct usbc_port_data *data = dev->data;
+
+	if (data->policy_cb_notify) {
+		data->policy_cb_notify(dev, notify);
+	}
+}
+
+/**
+ * @brief Notify the DPM of a WAIT message reception
+ */
+bool policy_wait_notify(const struct device *dev,
+			       const enum usbc_policy_wait_t notify)
+{
+	struct usbc_port_data *data = dev->data;
+
+	if (data->policy_cb_wait_notify) {
+		return data->policy_cb_wait_notify(dev, notify);
+	}
+
+	return false;
+}
+
+/**
+ * @brief Send the received source caps to the DPM
+ */
+void policy_set_src_cap(const struct device *dev,
+			       const uint32_t *pdos,
+			       const int num_pdos)
+{
+	struct usbc_port_data *data = dev->data;
+
+	if (data->policy_cb_set_src_cap) {
+		data->policy_cb_set_src_cap(dev, pdos, num_pdos);
+	}
+}
+
+/**
+ * @brief Get a Request Data Object from the DPM
+ */
+uint32_t policy_get_request_data_object(const struct device *dev)
+{
+	struct usbc_port_data *data = dev->data;
+
+	/* This callback must be implemented */
+	__ASSERT(data->policy_cb_get_rdo != NULL,
+		 "Callback pointer should not be NULL");
+
+	return data->policy_cb_get_rdo(dev);
+}
+
+/**
+ * @brief Check if the sink is a default level
+ */
+bool policy_is_snk_at_default(const struct device *dev)
+{
+	struct usbc_port_data *data = dev->data;
+
+	if (data->policy_cb_is_snk_at_default) {
+		return data->policy_cb_is_snk_at_default(dev);
+	}
+
+	return true;
+}
+
+/**
+ * @brief Get sink caps from the DPM
+ */
+void policy_get_snk_cap(const struct device *dev,
+			       uint32_t **pdos,
+			       int *num_pdos)
+{
+	struct usbc_port_data *data = dev->data;
+
+	/* This callback must be implemented */
+	__ASSERT(data->policy_cb_get_snk_cap != NULL,
+		 "Callback pointer should not be NULL");
+
+	data->policy_cb_get_snk_cap(dev, pdos, num_pdos);
+}
+
+static void pe_suspend_entry(void *obj)
+{
+	LOG_INF("PE_SUSPEND");
+}
+
+static void pe_suspend_run(void *obj)
+{
+	/* DO NOTHING */
+}
+
+/**
+ * @brief Policy engine State table
+ */
+static const struct smf_state pe_states[] = {
+	[PE_SNK_STARTUP] = SMF_CREATE_STATE(
```

Comment:
Right now, even though the state machine is split up, the two parts can't exist without each other. The structures of the state machine (constant state array, single state enum) mean that the C files can't be completely separated, and even if they could, there's no such thing as a port that is neither Source nor Sink.

I don't immediately see what this refactor buys us, but I'm guessing that the advantages will become more obvious when I review the source support PR.

Reply:
I thought it would be easier to review and maintain if the each file contained similar functionality, instead of having everything in one large file.

Reply:
In the long term, will all the files that make up the PE still get compiled together, or will some of them be left out based on the config?

Reply:
Yes. In a source only device, the pe_snk_* files will not be compiled in, and vice versa.

Reply:
Yes